onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -1,69 +1,87 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/common/common.h"
 #include "core/optimizer/constant_folding.h"
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/optimizer_execution_frame.h"
 #include "core/framework/op_kernel.h"
 #include "core/framework/ml_value.h"
 
+using namespace ::onnxruntime::common;
+
 namespace onnxruntime {
 
-Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified, bool& deleted) {
-  // Create execution frame for executing constant nodes.
-  OptimizerExecutionFrame::Info info({&node}, graph.GetAllInitializedTensors());
+Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-  std::vector<int> fetch_mlvalue_idxs;
-  for (const auto* node_out : node.OutputDefs()) {
-    fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-  }
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
```

Comment:
I don't think this check is needed... GetNode checks whether i is in range and throws exception if not. Regardless if you do want to check this I would suggest to add more descriptive message regarding the failure.

Reply:
In the current implementation of the order it is indeed not needed. I just wanted to be safe that we will never get a node index of a node that was somehow removed.